Whitelist and redirect domains

pezholio opened the pull request:
This is a common thing that comes out of pentests, so is worth having across all of our projects. I've also added an env var that allows for additional domains, if we want (for example if we transition from a temporary domain to a service.gov.uk domain, and want to old one to still work) 

I've also added a redirector which will redirect from a secondary domain to the canonical one with the path and query string intact.

pezholio committed:
Redirect requests not from the canonical hostname

erbridge commented on .env.example:
```diff
@@ -11,3 +11,11 @@ ROLLBAR_ENV=development
+# TODO: Replace `example.org` with a comma seperated list of any
+# other hostnames you want to app to respond to and redirect to the canonical
+# hostname, or delete this line completely
+ADDITIONAL_HOSTNAMES=example.org
```
Can we suppport an "empty" value of this and leave it blank, so:
```suggestion
ADDITIONAL_HOSTNAMES=
```

That way you don't have to remember how to put it back in when you actually release.

pezholio replied:
Done! 👍

erbridge commented on config/environments/production.rb:
```diff
@@ -93,6 +93,15 @@
 
+  # See https://github.com/rails/rails/issues/29893
+  # This whitelists the hosts the application can trust when using `url_for` and related helpers
+  if ENV["CANONICAL_DOMAIN"].present?
```
```suggestion
  if ENV["CANONICAL_HOSTNAME"].present?
```
?

Floppy replied:
oof, good spot, that completely passed me by.

pezholio replied:
Gaaaah - I initally had that var as `CANONICAL_DOMAIN`, but changed my mind 😃

erbridge commented on config/environments/production.rb:
```diff
@@ -93,6 +93,15 @@
   config.active_record.dump_schema_after_migration = false
 
+  # See https://github.com/rails/rails/issues/29893
+  # This whitelists the hosts the application can trust when using `url_for` and related helpers
```
Could we perhaps use "allow list" and "deny list" or similar instead of "whitelist" and "blacklist"?

pezholio replied:
Yep totally, thanks for the catch 👍

pezholio committed:
Only allowed hostnames can access the site

This is a common thing that comes out of pentests, so is worth having across all of
our projects.
